API documentation (#623)
anatawa12/AvatarOptimizer: .docs/content/docs/developers/make-your-tool-compatible-with-aao/index.md

your tool is already compatible with Avatar Optimizer!

If your tool adds some components to some portion of Avatar, your tool can be incompatible with Avatar Optimizer.

Since Avatar Optimizer has Garbage Collection system for Components and else, Avatar Optimizer have to 
know about all existing components in your Avatar at the optimization.

To avoid problem with unknown components, the Avatar Optimizer currently assumes unknown components
- have some side-effects.
- will have dependency relationship to all components referenced in the component.
  (They can be changed in the future.)

However, the assumption can be incorrect so Avatar Optimizer will generate the following warning.

![unknown-component-warning](unknown-component-warning.png)

If your tool is non-NDMF[^NDMF]-based non-destructive tools that will also be applied on entering play mode,
Avatar Optimizer might be proceed before applying your plugin.

## How to improve the compatibility? {#improve-compatibility}

### For NDMF based non-destructive tools {#improve-compatibility-ndmf-based}

If your tool is a non-destructive tools based on NDMF[^NDMF], please remove your components before

> Avatar Optimizer process. Avatar Optimizer does most thing in Optimization pass

Author: ここにおいてはbefore接続しのかたちのほうが自然だと思います

so if your plugin do nothing in Optimization pass, nothing is problem.
If your tool needs your components in Optimization pass, 
please execute before Avatar Optimizer with [`BeforePlugin`][ndmf-BeforePlugin]. 
QualifiedName of Avatar Optimizer in NDMF is `com.anatawa12.avatar-optimizer`.

If your tool actually want to do something with your components in Optimization pass,

> please [register your component][register-component] to Avatar Optimizer.

Reviewer: ```suggestion
please [register your components][register-component] to Avatar Optimizer.
```

### For non-NDMF based non-destructive tools {#improve-compatibility-non-ndmf-based}

If your tool is a non-destructive tools not based on NDMF[^NDMF], please consider
make your tool based on NDMF.

If your tool is applied on play, to ensure compatibility with Avatar Optimizer, you have to use NDMF to
guarantee applying ordering between Avatar Optimizer and your tool.
If your tool does something only on building avatar, making your tool based on NDMF is not required.

If you don't want to make your took based on NDMF, please remove your component before processing Avatar Optimizer.
To achieve this, please execute your tool before NDMF's Optimization pass.

> Currently NDMF executes Optimization passes in order `-1025`, JUST before VRCSDK's `RemoveAvatarEditorOnly` callback so

Author: Optimization phaseのpassは複数あるのでoptimization passesかoptimization phaseだと思う

> Currently NDMF executes Optimization passes in order `-1025`, JUST before VRCSDK's `RemoveAvatarEditorOnly` callback so

Reviewer: ふむ。まあそうですよね…
他の場所は単数が多かったと思いますが、そこは問題ない感じです?

> Currently NDMF executes Optimization passes in order `-1025`, JUST before VRCSDK's `RemoveAvatarEditorOnly` callback so

Author: どっちかというとNDMFのOptimiationを指してるの全部pass -> phaseにしたほうが良さそうなのでまとめてやります

your tool should register `IVRCSDKPreprocessAvatarCallback` with smaller `callbackOrder`.

If your tool actually want to do something with your components after Avatar Optimizer (Optimization pass in NDMF),

> please [register your component][register-component] to Avatar Optimizer.

Reviewer: ```suggestion
please [register your components][register-component] to Avatar Optimizer.
```

> ### For other tools that just holds data with components. {#non-destructive-tools}

Author: たしかにnon-いらないね

If your tool holds some information with components and doesn't have meaning at the build time, 
please remove your component before Avatar Optimizer with `IVRCSDKPreprocessAvatarCallback` or 

> register your component to Avatar Optimizer.

Reviewer: ```suggestion
register your components to Avatar Optimizer.
```

> When you want to remove your component with `IVRCSDKPreprocessAvatarCallback`, please refer [this section](#improve-compatibility-non-ndmf-based).

Reviewer: ```suggestion
If you want to remove your components with `IVRCSDKPreprocessAvatarCallback`, please refer [this section](#improve-compatibility-non-ndmf-based).
```

Reviewer: そもそもこの部分を無くして、70行目の with IVRC～あたりに`#improve-compatibility-non-ndmf-based`を付けたほうが良いような感じもします。

Author: まとめました

> When you want to register your component to Avatar Optimizer, please refer [this section][register-component].

Reviewer: ```suggestion
If you want to register your components to Avatar Optimizer, please refer [this section][register-component].
```

Reviewer: `#register-component`の話は、他の部分と同様に71行目に統合しても良さそうです。(ここは無くす)

Author: まとめた

### Registering your components {#register-component}

> If your tool want to keep your component after processing Avatar Optimizer, or want to removed by Avatar Optimizer,

Reviewer: ```suggestion
If your tool wants to keep your components after Avatar Optimizer process, or not to be removed by Avatar Optimizer,
```
後者、notが抜けているような気がしました

Author: 後者は実際にAAOに削除されてもいいものを指してるので`or to be removed`で正しいです

Reviewer: ふむ
うーん…

Reviewer: なるほど、理解しました
微妙に読みづらい気はしましたが、前後入れ替えも微妙ですし…

> you can register your component to Avatar Optimizer to tell about your component.

Reviewer: ```suggestion
you can register your components to Avatar Optimizer to tell about your components.
```

> To call APIs in Avatar Optimizer, first, Please make assembly definition file[^asmdef] if your tool doesn't have.

Reviewer: ```suggestion
First, to call APIs of Avatar Optimizer, please make an assembly definition file[^asmdef] if your tool doesn't have.
```

> Next, add `com.anatawa12.avatar-optimizer.api.editor` to assembly references in asmdef file.

Reviewer: ```suggestion
Next, add `com.anatawa12.avatar-optimizer.api.editor` to assembly references in asmdef file.  
```

If your tool doesn't want to depends on Avatar Optimizer, please use [Version Defines].

> Because Avatar Optimizer didn't have public API piror to 1.6.0 and will break api in 2.0.0, 

Reviewer: ```suggestion
Because Avatar Optimizer doesn't have public API prior to 1.6.0 and will break in 2.0.0, 
```

Author: 過去の話だからdidn't haveじゃない？

Reviewer: 過去かと言われると微妙な気が少ししました(今も旧バージョンは存在するので)

Author: ドキュメントの対象バージョンから相対的に見たら過去だと思う

> it's recommended to add version range like `[1.6,2.0)` (or more stricter like `[1.7,2.0)`).

Reviewer: 1.7以上にした方が良い理由があるのでしょうか…? (厳密性のためになる理由が掴めていません)

Author: 将来のAAOでAPI追加したときに1.7以上にしたくなる事があるので、それでもいいよって書いてるだけ。

Reviewer: あー

Reviewer: その意図が現在の表現では伝わりづらいように感じています。
```suggestion
it's recommended to add version range like `[1.6,2.0)` (or stricter like `[1.7,2.0)` when following new APIs that can be available in the future).
```
例えばこんな感じにするのはどうでしょう?

Author: 良さそうなので少し変えて適用しました

![version-defines.png](version-defines.png)

> Then, define `CompoinentInformation` for your component in your assembly.

Reviewer: ```suggestion
Then, define `ComponentInformation` for your component in your assembly.
```

```csharp
#if AVATAR_OPTIMIZER && UNITY_EDITOR

[ComponentInformation(typeof(YourComponent))]
internal class YourComponentInformation : ComponentInformation<YourComponent>
{
    protected override void CollectMutations(YourComponent component, ComponentMutationsCollector collector)
    {
        // call methods on the collector to tell about the component
    }

    protected override void CollectDependency(YourComponent component, ComponentDependencyCollector collector)
    {
        // call methods on the collector to tell about the component
    }
}

#endif
```

> In `CollectDependency`, you should register build-time or run-time dependencies of your component.

Reviewer: ```suggestion
In `CollectDependency`, you should register build-time or run-time dependencies of your component.  
```

> In `CollectMutations`, you should register any mutation your component may do.

Reviewer: ```suggestion
In `CollectMutations`, you should register any mutation your component may do.  
```

Please refer xmldoc and method name for more datails.

> If your component is just for keeping data for your in-editor tools, both will be empty method.

Reviewer: ```suggestion
If your component is just for keeping data for your in-editor tool, both will be empty method.
```

[fediverse]: https://misskey.niri.la/@anatawa12
[ndmf-BeforePlugin]: https://ndmf.nadena.dev/api/nadena.dev.ndmf.fluent.Sequence.html#nadena_dev_ndmf_fluent_Sequence_BeforePlugin_System_String_System_String_System_Int32_
[register-component]: #register-component

[^asmdef]: The file defines assembly other than Assembly-CSharp. Please refer [unity docs](https://docs.unity3d.com/2019.4/Documentation/Manual/ScriptCompilationAssemblyDefinitionFiles.html).

> [^NDMF]: [NDMF], Non-Destructive Modular Framework, is a framework for running non-destructive build plugins when 

Reviewer: ```suggestion
[^NDMF]: [NDMF], Non-Destructive Modular Framework, is a framework for executing non-destructive build plugins when 
```
原文がどこかにあったらすみません。他の文ではexecuteなので揃えました。

Author: 一応元は https://ndmf.nadena.dev/

This page describes the following two things.

- When can a tool be incompatible with Avatar Optimizer?
- How to improve the compatibility?

If you have some question, please feel free to ask [`@anatawa12@misskey.niri.la` on fediverse][fediverse].

## When can a tool be incompatible with Avatar Optimizer? {#when-incompatible}

If your tool doesn't add any components to the avatar and does nothing on the build time,
your tool is already compatible with Avatar Optimizer!

If your tool adds some components to portions of the avatar, your tool can be incompatible with Avatar Optimizer.

> Since Avatar Optimizer has Garbage Collection system for Components and others, Avatar Optimizer have to 

Reviewer: ```suggestion
Since Avatar Optimizer has Garbage Collection system for Components and others, Avatar Optimizer has to 
```
前reviewの適用忘れ?

so if your plugin do nothing in Optimization phase, nothing is problem.\
If your tool needs your components in Optimization phase, 
please execute before Avatar Optimizer processes with [`BeforePlugin`][ndmf-BeforePlugin]. 
QualifiedName of Avatar Optimizer in NDMF is `com.anatawa12.avatar-optimizer`.

If your tool actually wants to do something with your components in Optimization phase,
please [register your components][register-component] to Avatar Optimizer.

### For non-NDMF based non-destructive tools {#improve-compatibility-non-ndmf-based}

If your tool is a non-destructive tool not based on NDMF[^NDMF], please consider
make your tool based on NDMF.

If your tool is applied on play, to ensure compatibility with Avatar Optimizer, you have to use NDMF to
guarantee applying ordering between Avatar Optimizer and your tool.
If your tool does something only on building avatar, making your tool based on NDMF is not required.

If you don't want to make your tool based on NDMF, please remove your components before Avatar Optimizer processes.
To achieve this, your tool needs to execute before NDMF's Optimization phase.\

> Currently NDMF executes Optimization phase in order `-1025`, which is JUST before VRCSDK's `RemoveAvatarEditorOnly`

Reviewer: ```suggestion
Currently, NDMF executes Optimization phase in order `-1025`, which is JUST before VRCSDK's `RemoveAvatarEditorOnly`
```
前reviewの適用忘れか、意図的か…?
NDMFに掛けていたとすれば、"Current NDMF"の方が良さそうです。

Author: Current NDMFにしますね

callback, so your tool should register `IVRCSDKPreprocessAvatarCallback` with smaller `callbackOrder`.

If your tool actually wants to do something with your components after Avatar Optimizer processes 
(Optimization phase of NDMF), please [register your components][register-component] to Avatar Optimizer.

### For other tools that just hold data with components. {#improve-compatibility-destructive-tools}

> If your tool holds some information with components and have no meaning on the build time, 

Reviewer: ```suggestion
If your tool holds some information with components and has no meaning on the build time, 
```
前reviewの適用忘れ?

Next, add `com.anatawa12.avatar-optimizer.api.editor` to assembly references in asmdef file.\
If your tool doesn't want to depends on Avatar Optimizer, please use [Version Defines].
Because Avatar Optimizer didn't have public API prior to 1.6.0 and will break in 2.0.0, 
it's recommended to add version range like `[1.6,2.0)`

> (or stricter like `[1.7,2.0)` when you requires new APIs can be available in the future).

Reviewer: whichかthatは欲しい気がしました
```suggestion
(or stricter like `[1.7,2.0)` when you require new APIs which can be available in the future).
```
```suggestion
(or stricter like `[1.7,2.0)` when you require new APIs that can be available in the future).
```
